calendar-webadmin/src/main/java/com/linagora/calendar/webadmin/ResourceRoutes.java

```diff
@@ -156,84 +174,85 @@ private ResourceDTO getResource(Request req) {
+
+        return Mono.zip(creatorIdMono, domainIdMono, adminMapMono)
+            .flatMap(tuple -> {
+                OpenPaaSId creatorId = tuple.getT1();
+                OpenPaaSId domainId = tuple.getT2();
+                Map<Username, ResourceAdministrator> adminMap = tuple.getT3();
+                return resourceDAO.insert(buildInsertRequest(adminMap, creatorId, creationDTO, domainId))
+                    .flatMap(resourceId -> resourceAdministratorService.setAdmins(domainId, resourceId, adminMap.keySet().stream().toList())
+                        .thenReturn(resourceId));
+            })
+            .map(resourceId -> {
+                res.header(HttpHeader.LOCATION.asString(), getBasePath() + "/" + resourceId.value());
+                res.status(HttpStatus.CREATED_201);
+                return StringUtils.EMPTY;
+            })
             .block();
+    }
 
-        res.header("Location", getBasePath() + "/" + resourceId.value());
-        res.status(201);
-        return "";
+    private ResourceInsertRequest buildInsertRequest(Map<Username, ResourceAdministrator> adminMap, OpenPaaSId creatorId,
+                                                     ResourceCreationDTO creationDTO, OpenPaaSId domainId) {
+        return new ResourceInsertRequest(adminMap.values().stream().toList(), creatorId, creationDTO.description, domainId, creationDTO.icon, creationDTO.name);
     }
 
     private String updateResource(Request req, Response res) throws JsonExtractException {
         ResourceUpdateDTO dto = updateDTOJsonExtractor.parse(req.body());
-
-        ResourceId id = new ResourceId(req.params("id"));
-        Optional<List<ResourceAdministrator>> administrators = dto.administrators.map(this::retrieveResourceAdministrators);
-
-        try {
-            resourceDAO.update(id, new ResourceUpdateRequest(dto.name, dto.description, dto.icon, administrators))
-                .block();
-
-            res.status(204);
-            return "";
-        } catch (ResourceNotFoundException e) {
-            throw ErrorResponder.builder()
-                .statusCode(HttpStatus.NOT_FOUND_404)
-                .type(ErrorResponder.ErrorType.NOT_FOUND)
-                .message("Resource do not exist")
-                .cause(new RuntimeException())
-                .haltError();
-        }
+        ResourceId resourceId = new ResourceId(req.params("id"));
+
+        return findNotDeletedResource(resourceId)
+            .flatMap(resource -> {
+                Mono<Optional<List<ResourceAdministrator>>> newAdminsMono = dto.administrators()
+                    .map(admins -> resourceAdministratorService.updateAdmins(resource, admins).map(Optional::of))
+                    .orElse(Mono.just(Optional.empty()));
+
+                return newAdminsMono.flatMap(newAdminsOpt -> resourceDAO.update(resourceId,
+                    new ResourceUpdateRequest(dto.name(), dto.description(), dto.icon(), newAdminsOpt)));
+            })
```

Comment:
```suggestion
            .flatMap(resource ->  dto.administrators()
                    .map(admins -> resourceAdministratorService.updateAdmins(resource, admins).map(Optional::of))
                    .orElse(Mono.just(Optional.empty()))
                .map(newAdminsOpt -> new ResourceUpdateRequest(dto.name(), dto.description(), dto.icon(), newAdminsOpt))
                .flatMap(updateRequest -> resourceDAO.update(resourceId, updateRequest)))
```

We need smaller steps:

-> Split the flatMap in 2
-> extract the creation of the UpdateRequest

Thoughts?